File: perf_test/HashVsRBT/HashVsRBTTest.cpp
#include "HashVsRBTTest.h"
#include "HashTable.h"
#include "RBT.h"
#include "..\Utils\Utils.h"
#include "..\Utils\Guid.h"
#include <iostream>

using namespace std;

void HashVsRBT::HashVsRBTTest::DoTest()
{
	HashVsRBTTest t;
	t.HashTimeTest();
	t.RBTTimeTest();

	t.HashFindTest();
	t.RBTFindTest();
}

void HashVsRBT::HashVsRBTTest::HashTimeTest()
{
	cout << __FUNCTION__ << endl;
	auto time = Utils::MeasureTime([&]() {HashTable::CreateN(CaseSize); });
	cout << "time: " << time << "ms." << endl;
}

void HashVsRBT::HashVsRBTTest::RBTTimeTest()
{
	cout << __FUNCTION__ << endl;
	auto time = Utils::MeasureTime([&]() {RBT::CreateN(CaseSize); });
	cout << "time: " << time << "ms." << endl;
}

void HashVsRBT::HashVsRBTTest::HashFindTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = HashTable::CreateN(CaseSize);
	auto time = Utils::MeasureTime([&]() { 
		for (auto i = 0; i < CaseSize; ++i)
		{
			buffer.find(Utils::NewGuidString());
		}
	});
	cout << "time: " << time << "ms." << endl;
}

void HashVsRBT::HashVsRBTTest::RBTFindTest()
{
	cout << __FUNCTION__ << endl;
	auto buffer = RBT::CreateN(CaseSize);
	auto time = Utils::MeasureTime([&]() {
		for (auto i = 0; i < CaseSize; ++i)
		{
			buffer.find(Utils::NewGuidString());
		}
	});
	cout << "time: " << time << "ms." << endl;
}

File: src/PerfTest/HashVsRBT/HashVsRBTTest.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;

namespace PerfTest.HashVsRBT
{
    public class HashVsRBTTest
    {
        const int CaseSize = 200 * 1024;

        public static void DoTest()
        {
            var t = new HashVsRBTTest();

            t.HashCreateTest();
            t.RBTCreateTest();
            t.HashFindTest();
            t.RBTFindTest();
        }

        void HashCreateTest()
        {
            Console.WriteLine(nameof(HashCreateTest));
            var v = MeasureTime(() => HashRBTGenerator.CreateHashN(CaseSize));
            Console.WriteLine($"size: {v.Count}.");
        }

        void RBTCreateTest()
        {
            Console.WriteLine(nameof(RBTCreateTest));
            var v = MeasureTime(() => HashRBTGenerator.CreateRBTN(CaseSize));
            Console.WriteLine($"size: {v.Count}.");
        }

        void HashFindTest()
        {
            Console.WriteLine(nameof(HashFindTest));
            var buffer = HashRBTGenerator.CreateHashN(CaseSize);
            var v = MeasureTime(() =>
            {
                for (var i = 0; i < CaseSize; ++i)
                {
                    buffer.ContainsKey(Guid.NewGuid().ToString());
                }
                return 0;
            });
        }

        void RBTFindTest()
        {
            Console.WriteLine(nameof(RBTFindTest));
            var buffer = HashRBTGenerator.CreateRBTN(CaseSize);
            var v = MeasureTime(() =>
            {
                for (var i = 0; i < CaseSize; ++i)
                {
                    buffer.ContainsKey(Guid.NewGuid().ToString());
                }
                return 0;
            });
        }

        private T MeasureTime<T>(Func<T> action)
        {
            var sw = new Stopwatch();
            sw.Start();
            var t = action();
            Console.WriteLine($"{sw.ElapsedMilliseconds}ms.");
            return t;
        }
    }
}
